src/libsystemd-network/sd-dhcp-lease.c

```diff
@@ -559,6 +573,123 @@ static int lease_parse_sip_server(const uint8_t *option, size_t len, struct in_a
         return lease_parse_in_addrs(option + 1, len - 1, ret, n_ret);
 }
 
+static int lease_parse_dns_name(const uint8_t *optval, size_t optlen, char **ret) {
+        _cleanup_free_ char *name = NULL;
+        int r;
+
+        assert(optval);
+        assert(ret);
+
+        r = dns_name_from_wire_format(&optval, &optlen, &name);
+        if (r < 0)
+                return r;
+        if (r == 0 || optlen != 0)
+                return -EBADMSG;
+
+        *ret = TAKE_PTR(name);
+        return r;
+}
+
+static int lease_parse_dnr(const uint8_t *option, size_t len, sd_dns_resolver **dnr, size_t *n_dnr) {
+        int r;
+        sd_dns_resolver *res_list = NULL;
+        size_t n_resolvers = 0;
+        CLEANUP_ARRAY(res_list, n_resolvers, sd_dns_resolver_array_free);
+
+        assert(option);
+        assert(dnr);
+
+        _cleanup_(sd_dns_resolver_done) sd_dns_resolver res = {};
+
+        size_t offset = 0;
+        while (offset < len) {
+                if (offset + 2 > len)
+                        return -EBADMSG;
+
+                /* Instance Data length */
+                size_t ilen = unaligned_read_be16(&option[offset]);
+                if (offset + ilen + 2 > len)
+                        return -EBADMSG;
+                offset += 2;
+                size_t iend = offset + ilen;
+
+                res.priority = unaligned_read_be16(&option[offset]);
```

Comment:
Looks like it goes sideways here:
```sh
=================================================================
==49985==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x611000000275 at pc 0x00000055a346 bp 0x7ffd00027b50 sp 0x7ffd00027b48
READ of size 2 at 0x611000000275 thread T0
    #0 0x55a345 in unaligned_read_be16 /home/vagrant/systemd-asan/build/../src/basic/unaligned.h:14:27
    #1 0x55a345 in lease_parse_dnr /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-dhcp-lease.c:616:32
    #2 0x555712 in dhcp_lease_parse_options /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-dhcp-lease.c:1010:21
    #3 0x54a688 in parse_options /home/vagrant/systemd-asan/build/../src/libsystemd-network/dhcp-option.c:345:33
    #4 0x549efb in dhcp_option_parse /home/vagrant/systemd-asan/build/../src/libsystemd-network/dhcp-option.c:373:13
    #5 0x546652 in client_parse_message /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-dhcp-client.c:1477:13
    #6 0x53bbd2 in client_handle_offer_or_rapid_ack /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-dhcp-client.c:1562:13
    #7 0x53b5f4 in LLVMFuzzerTestOneInput /home/vagrant/systemd-asan/build/../src/libsystemd-network/fuzz-dhcp-client.c:78:16
    #8 0x442b44 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x442b44) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)
    #9 0x42ab17 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x42ab17) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)
    #10 0x4305d6 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x4305d6) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)
    #11 0x45cec6 in main (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x45cec6) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)
    #12 0x7f089f849b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #13 0x7f089f849c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #14 0x425124 in _start (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x425124) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)

0x611000000275 is located 0 bytes after 245-byte region [0x611000000180,0x611000000275)
allocated by thread T0 here:
    #0 0x4f7a52 in malloc (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x4f7a52) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)
    #1 0x7f08a0cb56cb in operator new(unsigned long) (/lib64/libstdc++.so.6+0xb56cb) (BuildId: 694f64a512ecd55e04f1587ddd790ec30cb0b726)
    #2 0x42ab17 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x42ab17) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)
    #3 0x4305d6 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x4305d6) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)
    #4 0x45cec6 in main (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x45cec6) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)
    #5 0x7f089f849b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #6 0x7f089f849c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #7 0x425124 in _start (/home/vagrant/systemd-asan/out/fuzz-dhcp-client+0x425124) (BuildId: c6d6134fbdd5fb5b91f37c9d874284e9fde1d4c7)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/vagrant/systemd-asan/build/../src/basic/unaligned.h:14:27 in unaligned_read_be16
Shadow bytes around the buggy address:
  0x610fffffff80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x611000000000: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x611000000080: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x611000000100: 00 00 00 00 00 00 05 fa fa fa fa fa fa fa fa fa
  0x611000000180: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x611000000200: 00 00 00 00 00 00 00 00 00 00 00 00 00 00[05]fa
  0x611000000280: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x611000000300: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x611000000380: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x611000000400: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x611000000480: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==49985==ABORTING
```

Reply:
It can be reproduced by running something like
```sh
cat <<'EOL' | xxd -ps -r >dnr-asan-overflow
07143a2a00000000000000230000f3a081a5000000000000000000000000
00000000000000006500001a000000000000000000000000000000000000
000000003004008200000102002d2b00ae01003801023004000200000230
00002b002f00feff48012b2b020100350102330400020001003701023304
000200000102ae0100c1010232040002000002002b00002f000200ae012b
2b00002f00ae012b3604d3fde9ffff777701b02f211f0c2d430801006000
01725c2e26357b1940c393686f0000003334000000000000000000007572
ffff0000ffff0026fc000010000100000000000000000000000000080065
a203000041
EOL
```
and passing `dnr-asan-overflow` to the fuzz target.

Reply:
The option was
```
          "dhcp.option.type": "162",
          "dhcp.option.type_tree": {
            "dhcp.option.length": "3",
            "dhcp.option.value": "00:00:41"
          }
```

Reply:
Oof, yeah I just accidentally over read in this case. Doesn't look like I repeated this error in dnrv6 though at least.